### ISSUE TITLE
torch.distributions.Binomial.sample() uses a massive amount of memory

### ISSUE THREAD

**Issue by jamestwebber** [pytorch/pytorch#20343]
## 🐛 Bug

I'd like to use a random sampling process as part of my training (related to work in [this paper](https://arxiv.org/abs/1901.11365). In my case this entails taking a binomial sample every epoch from the raw data, and using that sample to train. The dataset itself is small enough to fit into GPU memory (~8000 samples by 7000 features). Unfortunately, the binomial sample method contains [this](https://github.com/pytorch/pytorch/blob/master/torch/distributions/binomial.py#L97-L99):

```python
max_count = max(int(self.total_count.max()), 1)
shape = self._extended_shape(sample_shape) + (max_count,)
bernoullis = torch.bernoulli(self.probs.unsqueeze(-1).expand(shape))
```

Which, for my data, means that it tries to allocate a >600GB matrix to take a sample from a 240MB matrix. This is not ideal, to say the least. I guess no one has tried to sample a binomial distribution of this size before?

## To Reproduce

1. Create a binomial distribution with a big max count, e.g. `b = torch.distributions.binomial.Binomial(max_count=2000, probs=0.5)`
1. Try to take a big sample: `b.sample(sample_shape=torch.Size(10000, 10000)`
1. Laugh when pytorch tells you to buy 745GB of RAM just so you can perform this operation

## Expected behavior

The numpy version, on the same machine:

```
%timeit b = np.random.binomial(2000, 0.5, size=(10000, 10000))
11.4 s ± 60.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

## Environment

 - PyTorch Version (e.g., 1.0): 1.0.1.post2
 - OS (e.g., Linux): Ubuntu16 via Docker container
 - How you installed PyTorch (`conda`, `pip`, source): `conda install pytorch -c torch`
 - Build command you used (if compiling from source): N/A
 - Python version: 3.7
 - CUDA/cuDNN version: 9.0
 - GPU models and configuration: running on an NVIDIA Titan XP with 12GB of RAM
 - Any other relevant information: N/A?

**Comment by vishwakftw** [comment]
cc: @fritzo @neerajprad

**Comment by jamestwebber** [comment]
I will add that, while numpy appears to do this efficiently, [you might not want to emulate their implementation](https://github.com/numpy/numpy/blob/master/numpy/random/mtrand/distributions.c#L313)

**Comment by Lalaland** [comment]
For those large sizes couldn't you just use a normal approximation?

**Comment by jamestwebber** [comment]
That's a good point although I'm not it'll work in my case because I'm sampling from a heterogenous materix–the max count is high but most counts are low enough that a normal approximation wouldn't be appropriate. I would think that using the normal approximation on an item-by-item basis would be pretty hard.

Also, the code scales with the size of the sample as well as the max count, so even if the count is pretty low it'd run out of memory on large samples (to be fair, that might happen with any implementation).

Finally, it might be nice to speed this up in general. Running `Binomial(max_count=10, probs=0.5).sample(sample_shape=torch.Size((10000,10000)))` is pretty slow (like, 30 seconds on my machine) compared to ~9 seconds in numpy. I have no idea what they're doing in their code, though, it makes my eyes hurt to look at it.

**Comment by neerajprad** [comment]
Our binomial `.sample` method is naive, and relies on the native `bernoulli` sampler. This was put in place to provide a simple though naive implementation at the time, but given more interest and use from the community, I think it is worth considering implementing a native binomial sampler in Aten. 

scipy uses an algorithm called [BTPE](https://dl.acm.org/citation.cfm?id=42381), and the [tensorflow implementation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/random_binomial_op.cc), uses a more efficient rejection method called [BTRS](https://core.ac.uk/download/pdf/11007254.pdf). The BTRS paper also contains a comparison of these different samplers. If we do decide to implement this in Aten, we will also need to have a CUDA version of the sampler. Thanks for opening this up for discussion / contributions from the community.

**Comment by mbabadi** [comment]
I would like to add that I've run into a similar issue and I would appreciate a Aten native solution.

My inelegant and approximate solution at the moment is to truncate the support of the Binomial distribution around its mode, i.e. `|k - np| < MAX_SUPPORT`, calculate the Binomial `log_prob` for every allowed value, sample from a Categorical, and finally gather to Binomial. Of course this procedure is approximate (i.e. does not give you samples from `|k - np| >= MAX_SUPPORT`) and somewhat slower than the memory-hungry Bernoulli implementation, but at least the memory usage is bounded and is independent of `total_counts`.

**Comment by tpet93** [comment]
For someone having this issue, this may be helpful:

I was having the ram issue describe above when calling

        tensor.size()= [10,14,572,572]
        probability = [0.4]

        bino = torch.distributions.binomial.Binomial(tensor, probability)
        thinned = bino.sample()

        GPU RAM requested 45 GB
I got around the issue by looping through the first dimension of my tensor 10 times and only doing the binomial on a tensor of shape [14,572,572] and appending the output to get back to original shape. The looped version runs in 0.07s on a Nvidia P100,  while the original in numpy took 7s on a CPU

**Comment by jamestwebber** [comment]
Nice! I ended up doing the sampling I needed outside of PyTorch and then converting things, but it's definitely slower. I'm not sure how small I'd have to chunks things in my case, because the count array has a really wide range of values and so it might get very large in some cases.

So on that note, I've started working on adding BTRS to ATen directly. The CPU version was fairly straightforward (I haven't tested it yet but I'm on my way). I used `sample_poisson` and `_s_poisson_cpu` as templates and then ported the TensorFlow implementation over. If I get fancy maybe I'll try to implement BTRD which is supposed to be slightly better.

I'm less sure how to write the CUDA version. If someone more familiar with `ATen/native/cuda/Distributions.cu` is willing to give me guidance, I will write this up as a PR.

**Comment by jamestwebber** [comment]
Following up on that...I started adding to `Distributions.cpp` but now I see `sample_gamma` is defined in `Distributions.h` in a way that I guess works for both CPU and CUDA. Should I do that?

**Comment by jamestwebber** [comment]
^ I just made a PR with a BTRS implementation..the major hold-up was getting a development Docker image built so I could test on a GPU-enabled machine. Everything _seems_ to work, now, but we'll see what the test suite says.

**Comment by mbabadi** [comment]
@jamestwebber this is very exciting, thank you!! I very much look forward to this PR being polished and merged.

**Comment by mbabadi** [comment]
@jamestwebber do you get a chance to test your implementation?

**Comment by jamestwebber** [comment]
Yes, I tested it, things worked on my end at least. I don't know if the PR fell by the wayside or if there is some internal work happening that I can't see. You could maybe poke them on #31278
